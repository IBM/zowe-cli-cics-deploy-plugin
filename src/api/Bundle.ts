/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

"use strict";

import { Manifest } from "./Manifest";
import { BundlePart, IBundlePartDataType } from "./BundlePart";
import { NodejsappBundlePart } from "./NodejsappBundlePart";

/**
 * Class to represent a CICS Bundle.
 *
 * @export
 * @class Bundle
 */
export class Bundle {

  private path = require("path");
  private manifest: Manifest;
  private definedParts: BundlePart[] = [];
  private bundleDirectory: string;

  /**
   * Constructor for creating a Bundle.
   * @static
   * @param {string} directory - The bundle directory.
   * @throws ImperativeError
   * @memberof Bundle
   */
  constructor(directory: string) {
    this.bundleDirectory = this.path.normalize(directory);
    this.manifest = new Manifest(this.bundleDirectory);
  }

  /**
   * Returns the Bundle's identity (id) value.
   * @returns {string}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getId(): string {
    return this.manifest.getBundleId();
  }

  /**
   * Set the Bundle's identity (id) value.
   * @param {string} id - The identiry value for the Bundle.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public setId(id: string) {
    this.manifest.setBundleId(id);
  }

  /**
   * Set the Bundle's version number.
   * @param {number} majorVersion - The major version number.
   * @param {number} minorVersion - The minor version number.
   * @param {number} microVersion - The micro version number.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public setVersion(majorVersion: number, minorVersion: number, microVersion: number) {
    this.manifest.setBundleVersion(majorVersion, minorVersion, microVersion);
  }

  /**
   * Returns the Manifest contents as a JSON Object.
   *
   * @returns {object}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getManifest(): object {
    return this.manifest.getJson();
  }

  /**
   * Returns the Manifest contents as an XML string.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getManifestXML(): string {
    return this.manifest.getXML();
  }

  /**
   * Add a resource definition to the Bundle. If a part of the same name and type
   * already exists then it is replaced with the new part data.
   *
   * @param {IBundlePartDataType} partData - the part to add.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public addDefinition(partData: IBundlePartDataType) {
    // Create a BundlePart
    const bp = new BundlePart(this.bundleDirectory, partData, true, undefined);
    this.definedParts.push(bp);
    this.manifest.addDefinition(bp);
  }

  /**
   * Add a NODEJSAPP resource definition to the Bundle. If a NODEJSAPP of the same name
   * already exists then it is replaced with the new part data.
   *
   * @param {string} name - the name of the NODEJSAPP to add.
   * @param {string} startscript - the path to the starting script for the NODEJSAPP.
   * @param {number} port - the port mumber to associate with the NODEJSAPP.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public addNodejsappDefinition(name: string, startscript: string, port: number) {
    const nj = new NodejsappBundlePart(this.bundleDirectory, name, startscript, port);
    this.manifest.addDefinition(nj);
    this.definedParts.push(nj);
  }

  /**
   * Save the current Bundle. Any changes that have been made will be persisted.
   *
   * @throws ImperativeError
   * @memberof Bundle
   */
  public save() {
    this.definedParts.forEach( (value) => {
      value.save();
      });
    this.manifest.save();
  }
}
