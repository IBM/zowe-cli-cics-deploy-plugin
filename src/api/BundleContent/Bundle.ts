/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

"use strict";

import { Manifest } from "./Manifest";
import { BundlePart, IBundlePartDataType } from "./BundlePart";
import { NodejsappBundlePart } from "./NodejsappBundlePart";
import { IHandlerParameters } from "@zowe/imperative";

/**
 * Class to represent a CICS Bundle.
 *
 * @export
 * @class Bundle
 */
export class Bundle {

  /**
   * Get a template version of the .zosattributes file
   * @returns {string}
   * @memberof Bundle
   */
  public static getTemplateZosAttributesFile(): string {
    return `#z/OS File Attributes Document
#-----------------------------
# This document specfies the encodings for the files within
# the project in a form that is compatible with the
# 'Zowe files upload dir-to-uss' command.
#
# Don't upload node_modules
node_modules -
# Don't upload things that start with dots
.* -

# Upload the following file types in binary
*.jpg binary binary
*.png binary binary
*.gif binary binary
*.zip binary binary
*.eot binary binary
*.svg binary binary
*.ttf binary binary
*.woff binary binary

# Convert CICS Node.js profiles to EBCDIC
*.profile ISO8859-1 IBM-1047`;
  }

  private path = require("path");
  private fs = require("fs");
  private manifest: Manifest;
  private definedParts: BundlePart[] = [];
  private bundleDirectory: string;
  private merge: boolean;
  private overwrite: boolean;
  private preparedToSave: boolean = false;
  private zosAttribsNeeded: boolean = false;
  private zosAttribsAlreadyExists: boolean = false;
  private zosAttribsFile: string;
  private params: IHandlerParameters;

  /**
   * Constructor for creating a Bundle.
   * @static
   * @param {string} directory - The bundle directory.
   * @param {boolean} merge - Changes to the bundle manifest should be merged into any existing manifest.
   * @param {boolean} overwrite - Changes to the bundle contents should replace any existing contents.
   * @param {IHandlerParameters} params - The current Imperative handler parameters
   * @throws ImperativeError
   * @memberof Bundle
   */
  constructor(directory: string, merge: boolean, overwrite: boolean, params?: IHandlerParameters) {
    this.merge = merge;
    this.overwrite = overwrite;
    this.params = params;
    this.bundleDirectory = this.path.normalize(directory);
    this.manifest = new Manifest(this.bundleDirectory, this.merge, this.overwrite, params);
    this.preparedToSave = false;
    this.zosAttribsFile = this.path.join(this.bundleDirectory, ".zosattributes");
    this.zosAttribsNeeded = true;
  }

  /**
   * Validates that a Bundle appears to be valid on the file-system
   * @returns {string}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public validate() {

    // check that the manifest file exists. We could perform a more rigorous
    // validation of the contents of the Bundle, but this simple check is
    // sufficient to ensure the bundle is broadly valid on the file-system.
    // Errors will be thrown to report problems.
    this.manifest.validate();
  }

  /**
   * Determines whether the Bundle contains the named resource
   * @param {string} resource - The resource to query
   * @returns {boolean}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public contains(resource: string): boolean {
    const fullyQualifiedResourceName = this.path.join(this.bundleDirectory, resource);
    const relative = BundlePart.getRelativeFileReference(fullyQualifiedResourceName, this.bundleDirectory);
    if (relative === undefined) {
      // The file isn't within the Bundle directory
      return false;
    }

    const fullLocation = this.path.join(this.bundleDirectory, relative);
    if (!this.fs.existsSync(fullLocation)) {
      return false;
    }

    return true;
  }

  /**
   * Determines whether the Bundle contains any resources of the specified type
   * @param {string} resource - The resource to query
   * @returns {boolean}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public containsDefinitionsOfType(resourceType: string): boolean {
    return this.manifest.containsDefinitionsOfType(resourceType);
  }

  /**
   * Returns the Bundle's identity (id) value.
   * @returns {string}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getId(): string {
    return this.manifest.getBundleId();
  }

  /**
   * Set the Bundle's identity (id) value.
   * @param {string} id - The identity value for the Bundle.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public setId(id: string) {
    this.manifest.setBundleId(id);
  }

  /**
   * Returns the Bundle's version value.
   * @returns {string}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getVersion(): string {
    return this.manifest.getBundleVersion();
  }

  /**
   * Set the Bundle's version number.
   * @param {number} majorVersion - The major version number.
   * @param {number} minorVersion - The minor version number.
   * @param {number} microVersion - The micro version number.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public setVersion(majorVersion: number, minorVersion: number, microVersion: number) {
    this.manifest.setBundleVersion(majorVersion, minorVersion, microVersion);
  }

  /**
   * Returns the Manifest contents as a JSON Object.
   *
   * @returns {object}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getManifest(): object {
    return this.manifest.getJson();
  }

  /**
   * Returns the Manifest contents as an XML string.
   *
   * @returns {string}
   * @throws ImperativeError
   * @memberof Bundle
   */
  public getManifestXML(): string {
    return this.manifest.getXML();
  }

  /**
   * Add a resource definition to the Bundle. If a part of the same name and type
   * already exists then it is replaced with the new part data.
   *
   * @param {IBundlePartDataType} partData - the part to add.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public addDefinition(partData: IBundlePartDataType) {
    // Create a BundlePart
    this.preparedToSave = false;
    const bp = new BundlePart(this.bundleDirectory, partData, true, undefined, this.overwrite, this.params);
    this.definedParts.push(bp);
    this.manifest.addDefinition(bp);
  }

  /**
   * Add a NODEJSAPP resource definition to the Bundle. If a NODEJSAPP of the same name
   * already exists then it is replaced with the new part data.
   *
   * @param {string} name - the name of the NODEJSAPP to add.
   * @param {string} startscript - the path to the starting script for the NODEJSAPP.
   * @param {number} port - the port mumber to associate with the NODEJSAPP.
   * @throws ImperativeError
   * @memberof Bundle
   */
  public addNodejsappDefinition(name: string, startscript: string, port: number) {
    this.preparedToSave = false;
    const nj = new NodejsappBundlePart(this.bundleDirectory, name, startscript, port, this.overwrite, this.params);
    this.manifest.addDefinition(nj);
    this.definedParts.push(nj);
    this.zosAttribsNeeded = true;
  }

  /**
   * Each bundle part is prompted to ensure that it is in a fit state to be saved.
   * This might involve ensuring that file system permissions are suitable and that
   * existing files wont be overwritten. The goal is that errors can be detected
   * before any changes are actually saved, thereby reducing the probability that
   * something will go wrong after partial changes are saved, and potentially resulting
   * in a damaged Bundle on the file system.
   *
   * @throws ImperativeError
   * @memberof Bundle
   */
  public prepareForSave() {
    // prepare the manifest, this will check write access to the target directory
    // (amongst other things).
    this.manifest.prepareForSave();

    // prepare each bundle part in turn
    this.definedParts.forEach( (value) => {
      value.prepareForSave();
      });

    // prepare the .zosattributes file (which can be useful for transferring a
    // bundle to zFS).
    this.prepareZosAttribs();

    this.preparedToSave = true;
  }

  /**
   * Save the current Bundle. Any changes that have been made will be persisted.
   *
   * @throws ImperativeError
   * @memberof Bundle
   */
  public save() {
    // Prepare the resources to be saved, without actually saving them. This should
    // cause most common errors to be detected before we start persisting anything.
    // If an error does occur mid-save then we have a better chance of avoiding a
    // broken Bundle.
    if (this.preparedToSave === false) {
      this.prepareForSave();
    }

    // save the parts
    this.definedParts.forEach( (value) => {
      value.save();
      });

    // save the zosattributes
    this.writeZosAttribs();

    // save the manifest
    this.manifest.save();
  }

  private prepareZosAttribs() {
    if (this.zosAttribsNeeded) {
      this.zosAttribsAlreadyExists = BundlePart.alreadyExists(this.zosAttribsFile, this.overwrite);
    }
  }

  private writeZosAttribs() {
    if (!this.zosAttribsNeeded) {
      return;
    }

    const contents = Bundle.getTemplateZosAttributesFile();

    // Write the zosattributes file
    try {
      if (this.params !== undefined) {
        let action = "    create";
        if (this.zosAttribsAlreadyExists) {
          action = " overwrite";
        }
        this.params.response.console.log(action + " : " + this.path.relative(this.bundleDirectory, this.zosAttribsFile));
      }
      this.fs.writeFileSync(this.zosAttribsFile, contents, "utf8");
    }
    catch (err) {
      throw new Error("An error occurred attempting to write .zosattributes file '" + this.zosAttribsFile + "': " + err.message);
    }
  }
}
