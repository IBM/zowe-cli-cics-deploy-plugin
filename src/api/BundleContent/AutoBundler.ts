/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

"use strict";

import { Bundle } from "./Bundle";
import { Logger, IHandlerParameters } from "@zowe/imperative";


/**
 * Utility class to bundle enable the working directory. If a package.json file is
 * found then the contents are processed and a basic NODEJSAPP resource is added
 * to the Bundle.
 *
 * @export
 * @class AutoBundler
 */
export class AutoBundler {

  private MAX_NODEJSAPP_LEN = 32;
  private bundle: Bundle;
  private fs = require("fs");
  private path = require("path");
  private packageJsonFile: string;
  private bundleidOverride: string;
  private bundleMajorVerOverride: number;
  private bundleMinorVerOverride: number;
  private bundleMicroVerOverride: number;
  private nodejsappOverride: string;
  private startscriptOverride: string;
  private portOverride: number;
  private bundleDirectory: string;
  private overwrite: boolean = false;
  private merge: boolean = false;

  /**
   * Constructor to perform the automatic bundle enablement.
   * @param {string} directory - The bundle directory.
   * @param {IHandlerParameters} params - The Imperative handler parameters
   * @static
   * @throws ImperativeError
   * @memberof AutoBundler
   */
  constructor(directory: string, params: IHandlerParameters) {

    this.bundleDirectory = this.path.normalize(directory);
    this.packageJsonFile = directory + "/package.json";
    this.validateMergeAndOverwrite(params.arguments.merge, params.arguments.overwrite);
    this.bundle = new Bundle(this.bundleDirectory, this.merge, this.overwrite, params);
    this.discoverArguments(params);

    this.autoDetectWorkdirContent(params);
  }

  /**
   * Returns the generated Bundle instance.
   * @throws ImperativeError
   * @returns {Bundle}
   * @memberof AutoBundler
   */
  public getBundle(): Bundle {
    return this.bundle;
  }

  /**
   * Save the generated Bundle
   *
   * @throws ImperativeError
   * @memberof AutoBundler
   */
  public save() {
    this.bundle.save();
  }

  private autoDetectWorkdirContent(params: IHandlerParameters) {

    // Look for a package.json file; if found, process the contents for a NODEJSAPP
    if (this.fs.existsSync(this.packageJsonFile)) {
      this.processPackageJson(params);
    }
    else {
      // If we get here then we've not found a package.json file. it remains possible
      // that sufficient command line parameters have been set to allow a NODEJSAPP
      // to be created. If so, process them.
      if (this.startscriptOverride !== undefined ||
          this.nodejsappOverride   !== undefined ||
          this.portOverride        !== undefined ) {
        this.bundle.addNodejsappDefinition(this.nodejsappOverride, this.startscriptOverride, this.portOverride);
        try {
          const msg = 'define : NODEJSAPP "' + this.nodejsappOverride + '" with startscript "' + this.startscriptOverride + '"';
          params.response.console.log(msg);

          // Also log the message for posterity
          if (params.arguments.silent === undefined) {
            const logger = Logger.getAppLogger();
            logger.debug(msg);
          }
        }
        catch (error) {
          // logging errors can be thrown in some of the mocked tests... just ignore it.
        }
      }
    }
  }

  private processPackageJson(params: IHandlerParameters) {

    // read the package.json file
    let pj;
    try {
      pj = JSON.parse(this.fs.readFileSync(this.packageJsonFile, "utf8"));
    }
    catch (exception)
    {
      throw new Error("Parsing error occurred reading package.json: " + exception.message);
    }

    // Find the name from package.json
    const name = pj.name;

    // Set the name of the Bundle based on what was found
    if (this.bundleidOverride === undefined) {
      if (name === undefined) {
        throw new Error("No bundleid value set");
      }
      this.bundle.setId(name);
    }

    // Recover the mangled variant of the name (it may be different from what we started with)
    const bundleName = this.bundle.getId();

    // Set the Bundle version to 1.0.0
    if (this.bundleMajorVerOverride === undefined) {
      this.bundle.setVersion(1, 0, 0);
    }

    // truncate the Bundle name to find a suitable NODEJSAPP resource name
    let njappname = this.nodejsappOverride;
    if (njappname === undefined) {
      njappname = bundleName;
    }
    njappname = njappname.substring(0, this.MAX_NODEJSAPP_LEN);

    // get the start script
    let ss = this.startscriptOverride;
    let fullSS = ss;
    if (ss === undefined) {
      if (pj.scripts !== undefined) {
        ss = pj.scripts.start;
      }
      if (ss !== undefined) {
        // Remove the first word from the value (which will be the node command)
        ss = ss.substr(ss.indexOf(" ") + 1).trim();
        fullSS = this.bundleDirectory + "/" + ss;
      }
      // If there's no start script, try the main script instead
      else {
        ss = pj.main;

        if (ss !== undefined) {
         fullSS = this.bundleDirectory + "/" + pj.main;
        }
      }
    }

    // Add the NODEJSAPP to the Bundle
    this.bundle.addNodejsappDefinition(njappname, fullSS, this.portOverride);
    try {
      // Construct an info message to report that the NODEJSAPP has been created
      const msg = '    define : NODEJSAPP "' + njappname + '" with startscript "' + ss + '"';

      // log the message to the console for the uesr
      params.response.console.log(msg);

      // Also log the message for posterity
      if (params.arguments.silent === undefined) {
        const logger = Logger.getAppLogger();
        logger.debug(msg);
      }
    }
    catch (error) {
      // logging errors can be thrown in some of the mocked tests... just ignore it.
    }
  }

  private discoverArguments(params: IHandlerParameters) {

    const args = params.arguments;

    if (args.bundleid !== undefined) {
      this.validateBundleId(args.bundleid);
    }
    if (args.bundleversion !== undefined) {
      this.validateBundleVer(args.bundleversion);
    }
    if (args.nodejsapp !== undefined) {
      this.validateNodejsapp(args.nodejsapp);
    }
    if (args.startscript !== undefined) {
      this.validateStartscript(args.startscript);
    }
    if (args.port !== undefined) {
      this.validatePort(args.port);
    }
  }

  private validateBundleId(value: string) {
    this.bundleidOverride = value;
    this.bundle.setId(this.bundleidOverride);
  }

  private validateBundleVer(value: string) {
    const splitString = value.split(".");
    this.bundleMajorVerOverride = parseInt(splitString[0], 10);
    this.bundleMinorVerOverride = parseInt(splitString[1], 10);
    this.bundleMicroVerOverride = parseInt(splitString[2], 10);
    this.bundle.setVersion(this.bundleMajorVerOverride,
                           this.bundleMinorVerOverride,
                           this.bundleMicroVerOverride);
  }

  private validateNodejsapp(value: string) {
    this.nodejsappOverride = value;
  }

  private validateStartscript(value: string) {
    this.startscriptOverride = value;
  }

  private validatePort(value: string) {
    this.portOverride = parseInt(value, 10);
  }

  private validateMergeAndOverwrite(merge: boolean, overwrite: boolean) {
    if (merge === true && overwrite !== true) {
      throw new Error("--merge requires the use of --overwrite");
    }
    this.merge = merge;
    this.overwrite = overwrite;
  }
}
