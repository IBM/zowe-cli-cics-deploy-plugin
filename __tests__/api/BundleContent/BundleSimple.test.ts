/*
* This program and the accompanying materials are made available under the terms of the
* Eclipse Public License v2.0 which accompanies this distribution, and is available at
* https://www.eclipse.org/legal/epl-v20.html
*
* SPDX-License-Identifier: EPL-2.0
*
* Copyright IBM Corp, 2019
*
*/

import { Bundle } from "../../../src/api/BundleContent/Bundle";
import * as fs from "fs";

describe("Bundle01", () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });
    it("should read an existing bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", true, true);

        // Check the output as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
    });
    it("set the id and version", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", true, true);

        // Set the Id
        bund.setId("TestExample");
        expect(bund.getId()).toMatch("TestExample");

        // Set the Version
        // eslint-disable-next-line no-magic-numbers
        bund.setVersion(33, 44, 55);

        // Check the output as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
    });
    it("add definitions", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition
        bund.addDefinition({name: "name1", type: "type1", path: "__tests__/__resources__/ExampleBundle02/Artefact1"});
        bund.addDefinition({name: "name2", type: "type2", path: "__tests__/__resources__/ExampleBundle02/Artefact2.txt"});
        bund.addDefinition({name: "name3", type: "type3", path: "__tests__/__resources__/ExampleBundle02/folder/Artefact3.jpg"});

        // Check the output as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
    });
    it("add a definition that references files outside of the bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition that's out of scope
        let err: Error;
        try {
            bund.addDefinition({name: "name1", type: "type1", path: "__tests__/__resources__/ExampleBundle01/META-INF/cics.xml"});
        }
        catch (error) {
            err = error;
        }

        // Check the output as JSON
        expect(err).toBeDefined();
        expect(err.message).toContain("BundlePart \"name1\" references a file outside of the Bundle directory:");
        expect(err.message).toContain("cics.xml");
    });
    it("add a not found definition", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition
        let err: Error;
        try {
            bund.addDefinition({name: "name1", type: "type1", path: "__tests__/__resources__/ExampleBundle02/Artefact3"});
        }
        catch (error) {
            err = error;
        }

        // Check the output as JSON
        expect(err).toBeDefined();
        expect(err.message).toContain("BundlePart \"name1\" references a file that does not exist:");
        expect(err.message).toContain("Artefact3");
    });
    it("add a part with missing name", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition
        let err: Error;
        try {
            bund.addDefinition({name: undefined, type: "type1", path: "__tests__/__resources__/ExampleBundle02/Artefact1"});
        }
        catch (error) {
            err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("add a part with missing type", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition
        let err: Error;
        try {
            bund.addDefinition({name: "name1", type: undefined, path: "__tests__/__resources__/ExampleBundle02/Artefact1"});
        }
        catch (error) {
            err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("add a part with missing path", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition
        let err: Error;
        try {
            bund.addDefinition({name: "name1", type: "type1", path: undefined});
        }
        catch (error) {
            err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("tolerate an existing almost empty manifest", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle03", true, true);

        // Add a definition
        bund.addDefinition({name: "name1", type: "type1", path: "__tests__/__resources__/ExampleBundle03/Artefact1"});

        // Check the output as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
    });
    it("should be able to add a resource to a single item manifest", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle06", true, true);

        // Add a definition
        bund.addDefinition({name: "item2", type: "type2", path: "__tests__/__resources__/ExampleBundle06/Artefact1"});

        // Check the output as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
    });
    it("add a NODEJSAPP", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle03", true, true);

        // Add a definition
        // eslint-disable-next-line no-magic-numbers
        bund.addNodejsappDefinition("NodeName", "__tests__/__resources__/ExampleBundle03/Artefact1", 1000);

        // Check the manifest as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
        // Check the manifest as XML
        expect(bund.getManifestXML()).toMatchSnapshot();
    });
    it("should warn that an existing bundle cant be overwritten", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle01", false, false);

        let err: Error;
        try {
            bund.prepareForSave();
        }
        catch (error) {
            err = error;
        }

        // Check the output as JSON
        expect(err.message).toMatchSnapshot();
    });
    it("should be able to save new definitions", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        // Add a definition
        bund.addDefinition({name: "name1", type: "type1", path: "__tests__/__resources__/ExampleBundle02/Artefact1"});
        bund.prepareForSave();

        // Check the output as JSON
        expect(JSON.stringify(bund.getManifest())).toMatchSnapshot();
    });
    it("should find a directory within the Bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        expect(bund.contains("META-INF")).toBeTruthy();
    });
    it("should find a file within the Bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        expect(bund.contains("Artefact2.txt")).toBeTruthy();
    });
    it("should find a file within a directory within the Bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        expect(bund.contains("META-INF/cics.xml")).toBeTruthy();
    });
    it("should not find a missing file within the Bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        expect(bund.contains("doesNotexist")).toBeFalsy();
    });
    it("should not find a file outside of the Bundle", () => {

        // Create a Bundle
        const bund = new Bundle("__tests__/__resources__/ExampleBundle02", true, true);

        expect(bund.contains("../ExampleBundle01")).toBeFalsy();
    });
});
