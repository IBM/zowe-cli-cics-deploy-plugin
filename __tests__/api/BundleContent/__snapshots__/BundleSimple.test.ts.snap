// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bundle01 add a NODEJSAPP 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","define":[{"name":"NodeName","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/NodeName.nodejsapp"}]}}"`;

exports[`Bundle01 add a NODEJSAPP 2`] = `
"<manifest xmlns="http://www.ibm.com/xmlns/prod/cics/bundle"><define name="NodeName" type="http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP" path="nodejsapps/NodeName.nodejsapp"></define></manifest>
"
`;

exports[`Bundle01 add a part with missing name 1`] = `"BundlePart name is not set."`;

exports[`Bundle01 add a part with missing path 1`] = `"BundlePart path is not set for part "name1""`;

exports[`Bundle01 add a part with missing type 1`] = `"BundlePart type is not set for part "name1""`;

exports[`Bundle01 add definitions 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"ThisIsAnId","bundleMajorVer":"10","bundleMinorVer":"11","bundleMicroVer":"12","define":[{"name":"name1","type":"type1","path":"Artefact1"},{"name":"name2","type":"type2","path":"Artefact2.txt"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"},{"name":"name3","type":"type3","path":"folder/Artefact3.jpg"}]}}"`;

exports[`Bundle01 set the id and version 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"TestExample","bundleMajorVer":33,"bundleMinorVer":44,"bundleMicroVer":55,"define":[{"name":"name1","type":"type1","path":"path1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Bundle01 should be able to add a resource to a single item manifest 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","define":[{"name":"single","type":"type1","path":"path1"},{"name":"item2","type":"type2","path":"Artefact1"}]}}"`;

exports[`Bundle01 should be able to save new definitions 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"ThisIsAnId","bundleMajorVer":"10","bundleMinorVer":"11","bundleMicroVer":"12","define":[{"name":"name1","type":"type1","path":"Artefact1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Bundle01 should read an existing bundle 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","bundleVersion":"1","bundleRelease":"2","id":"ThisIsAnId","bundleMajorVer":"10","bundleMinorVer":"11","bundleMicroVer":"12","define":[{"name":"name1","type":"type1","path":"path1"},{"name":"name2","type":"type2","path":"path2"},{"name":"name3","type":"http://www.ibm.com/xmlns/prod/cics/bundle/NODEJSAPP","path":"nodejsapps/Test.nodejsapp"}]}}"`;

exports[`Bundle01 should warn that an existing bundle cant be overwritten 1`] = `"A bundle manifest file already exists. Specify --overwrite to replace it, or --merge to merge changes into it."`;

exports[`Bundle01 tolerate an existing almost empty manifest 1`] = `"{"manifest":{"xmlns":"http://www.ibm.com/xmlns/prod/cics/bundle","define":[{"name":"name1","type":"type1","path":"Artefact1"}]}}"`;
